swarm/src/connection.rs

```diff
@@ -42,13 +42,36 @@ use libp2p_core::connection::ConnectedPoint;
 use libp2p_core::multiaddr::Multiaddr;
 use libp2p_core::muxing::{StreamMuxerBox, StreamMuxerEvent, StreamMuxerExt, SubstreamBox};
 use libp2p_core::upgrade::{InboundUpgradeApply, OutboundUpgradeApply};
-use libp2p_core::PeerId;
 use libp2p_core::{upgrade, UpgradeError};
+use libp2p_core::{Endpoint, PeerId};
 use std::future::Future;
 use std::task::Waker;
 use std::time::Duration;
 use std::{fmt, io, mem, pin::Pin, task::Context, task::Poll};
 
+/// Connection identifier.
+#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct ConnectionId(usize);
+
+impl ConnectionId {
+    /// Creates a `ConnectionId` from a non-negative integer.
+    ///
+    /// This is primarily useful for creating connection IDs
+    /// in test environments. There is in general no guarantee
+    /// that all connection IDs are based on non-negative integers.
+    pub fn new(id: usize) -> Self {
+        Self(id)
+    }
+}
+
+impl std::ops::Add<usize> for ConnectionId {
+    type Output = Self;
+
+    fn add(self, other: usize) -> Self {
+        Self(self.0 + other)
+    }
+}
+
```

Comment:
Remind me, when do we add two `ConnectionId`s? Or is this just to get the next one? In the latter case, it would not make sense to implement the public `Add` trait, right? (Feel free to ignore in this pull request.)

Reply:
This will get removed in https://github.com/libp2p/rust-libp2p/pull/3327.

In this PR, I am just moving things around without touching functionality!